mrmustard/lab_dev/circuit_components.py

```diff
@@ -489,65 +470,70 @@ def to_fock(self, shape: int | Sequence[int] | None = None) -> CircuitComponent:
                 defaults to the value of ``AUTOSHAPE_MAX`` in the settings.
         """
-        fock = Fock(math.astensor(self.fock(shape, batched=True)), batched=True)
+        fock = Fock(self.fock(shape, batched=True), batched=True)
         fock._original_bargmann_data = self.representation.data
-        return self._from_attributes(fock, self.wires, self.name)
+        try:
+            ret = self._from_param_set(fock, self.wires, self.name)
+        except TypeError:
+            ret = self._from_attributes(fock, self.wires, self.name)
+        return ret
 
-    def auto_shape(self, **_) -> tuple[int, ...]:
+    def _add_parameter(self, parameter: Union[Constant, Variable]):
         r"""
-        The shape of the Fock representation of this component. If the component has a Fock representation
-        then it is just the shape of the array. If the components is a State in Bargmann
-        representation the shape is calculated using autoshape using the single-mode marginals.
-        If the component is not a State then the shape is a tuple of ``settings.AUTOSHAPE_MAX`` values
-        except where the ``manual_shape`` attribute has been set..
-        """
-        return tuple(s or settings.AUTOSHAPE_MAX for s in self.manual_shape)
+        Adds a parameter to this circuit component and makes it accessible as an attribute.
 
-    def __add__(self, other: CircuitComponent) -> CircuitComponent:
-        r"""
-        Implements the addition between circuit components.
-        """
-        if self.wires != other.wires:
-            raise ValueError("Cannot add components with different wires.")
-        rep = self.representation + other.representation
-        name = self.name if self.name == other.name else ""
-        return self._from_attributes(rep, self.wires, name)
+        Args:
+            parameter: The parameter to add.
 
-    def __sub__(self, other: CircuitComponent) -> CircuitComponent:
-        r"""
-        Implements the subtraction between circuit components.
+        Raises:
+            ValueError: If the the given parameter is incompatible with the number
+                of modes (e.g. for parallel gates).
         """
-        if self.wires != other.wires:
-            raise ValueError("Cannot subtract components with different wires.")
-        rep = self.representation - other.representation
-        name = self.name if self.name == other.name else ""
-        return self._from_attributes(rep, self.wires, name)
+        if parameter.value.shape != ():
+            length = sum(parameter.value.shape)
+            if length != 1 and length != len(self.modes):
+                raise ValueError(f"Length of ``{parameter.name}`` must be 1 or {len(self.modes)}.")
+        self.parameter_set.add_parameter(parameter)
+        self.__dict__[parameter.name] = parameter
 
-    def __mul__(self, other: Scalar) -> CircuitComponent:
+    def _from_param_set(
```

Comment:
I can't quite put my finger on it, but something feels weird about this. It's effectively just a copy, then the representation gets swapped out. we probably don't need `name` or `wires`, since those are handled by `self.__class__` and `self.modes` inherently.

would it make more sense to repurpose this work, and instead implement `__copy__` for `CircuitComponent`?

Reply:
Agreed. The main reason for separating it out this way was the hope that serialization might make use of it. Tbh the better way I think is to just move `_getitem_builtin_state` (the method I copied the logic from) to `CircuitComponent` as `_getitem_builtin`.

Reply:
I didn't know this existed!! that sounds pretty good to me if it can be generalized, but the problem I hit was that `self.modes` isn't _necessarily_ defined. In practice, it is. All concrete classes have it. Only intermediate abstract classes (`Operation`, `Map`, `Unitary` and `Channel`) are constructed differently (using `modes_in` and `modes_out`), as well as the base types (`CircuitComponent`, `Transformation` and `State`) which are constructed using `wires`. Evidently, we assume in `State` that `self.modes` is defined, so we can keep that assumption.

Reply:
Yeah, I agree this was something that bothered me also. I think since it is a hidden method if we just document that this is meant to only work on user facing classes it should be fine? Otherwise then I guess this is a case where we just have to have some duplicate code and I can keep it in `to_fock`